File: src/main/java/no/nav/bidrag/beregn/forskudd/rest/service/BeregnForskuddService.java
package no.nav.bidrag.beregn.forskudd.rest.service;

import static no.nav.bidrag.beregn.forskudd.rest.BidragBeregnForskudd.SECURE_LOGGER;

import com.fasterxml.jackson.databind.ObjectMapper;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.Collection;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.stream.Collectors;
import no.nav.bidrag.beregn.felles.dto.AvvikCore;
import no.nav.bidrag.beregn.forskudd.core.ForskuddCore;
import no.nav.bidrag.beregn.forskudd.core.dto.BeregnetForskuddResultatCore;
import no.nav.bidrag.beregn.forskudd.core.dto.ResultatPeriodeCore;
import no.nav.bidrag.beregn.forskudd.rest.consumer.SjablonConsumer;
import no.nav.bidrag.beregn.forskudd.rest.dto.http.BeregnForskuddGrunnlag;
import no.nav.bidrag.beregn.forskudd.rest.dto.http.BeregnetForskuddResultat;
import no.nav.bidrag.beregn.forskudd.rest.dto.http.ResultatGrunnlag;
import no.nav.bidrag.beregn.forskudd.rest.exception.UgyldigInputException;
import no.nav.bidrag.commons.web.HttpResponse;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;

@Service
public class BeregnForskuddService {

  private static final Logger LOGGER = LoggerFactory.getLogger(BeregnForskuddService.class);

  private final SjablonConsumer sjablonConsumer;
  private final ForskuddCore forskuddCore;

  public BeregnForskuddService(SjablonConsumer sjablonConsumer, ForskuddCore forskuddCore) {
    this.sjablonConsumer = sjablonConsumer;
    this.forskuddCore = forskuddCore;
  }

  public HttpResponse<BeregnetForskuddResultat> beregn(BeregnForskuddGrunnlag grunnlag) {

    // Kontroll av inputdata
    grunnlag.valider();

    // Henter sjabloner
    var sjablonSjablontallResponse = sjablonConsumer.hentSjablonSjablontall();
    if (LOGGER.isDebugEnabled() && sjablonSjablontallResponse.getResponseEntity().getBody() != null) {
      LOGGER.debug("Antall sjabloner hentet av type Sjablontall: {}", sjablonSjablontallResponse.getResponseEntity().getBody().size());
    }

    // Lager input-grunnlag til core-modulen
    var grunnlagTilCore = CoreMapper.mapGrunnlagTilCore(grunnlag, sjablonSjablontallResponse.getResponseEntity().getBody());

    BeregnetForskuddResultatCore resultatFraCore;

    // Kaller core-modulen for beregning av forskudd
    if (SECURE_LOGGER.isDebugEnabled()) {
      SECURE_LOGGER.debug("Forskudd - grunnlag for beregning: {}", grunnlagTilCore);
    }

    try {
      resultatFraCore = forskuddCore.beregnForskudd(grunnlagTilCore);
    } catch (Exception e) {
      throw new UgyldigInputException("Ugyldig input ved beregning av forskudd: " + e.getMessage());
    }

    if (!resultatFraCore.getAvvikListe().isEmpty()) {
      LOGGER.error("Ugyldig input ved beregning av forskudd. Følgende avvik ble funnet: " + System.lineSeparator()
          + resultatFraCore.getAvvikListe().stream().map(AvvikCore::getAvvikTekst).collect(Collectors.joining(System.lineSeparator())));
      SECURE_LOGGER.error("Ugyldig input ved beregning av forskudd. Følgende avvik ble funnet: " + System.lineSeparator()
          + resultatFraCore.getAvvikListe().stream().map(AvvikCore::getAvvikTekst).collect(Collectors.joining(System.lineSeparator())));
      SECURE_LOGGER.info("Forskudd - grunnlag for beregning: " + System.lineSeparator()
          + "beregnDatoFra= " + grunnlagTilCore.getBeregnDatoFra() + System.lineSeparator()
          + "beregnDatoTil= " + grunnlagTilCore.getBeregnDatoTil() + System.lineSeparator()
          + "soknadBarn= " + grunnlagTilCore.getSoknadBarn() + System.lineSeparator()
          + "barnIHusstandenPeriodeListe= " + grunnlagTilCore.getBarnIHusstandenPeriodeListe() + System.lineSeparator()
          + "inntektPeriodeListe= " + grunnlagTilCore.getInntektPeriodeListe() + System.lineSeparator()
          + "sivilstandPeriodeListe= " + grunnlagTilCore.getSivilstandPeriodeListe() + System.lineSeparator());
      throw new UgyldigInputException("Ugyldig input ved beregning av forskudd. Følgende avvik ble funnet: "
          + resultatFraCore.getAvvikListe().stream().map(AvvikCore::getAvvikTekst).collect(Collectors.joining("; ")));
    }

    if (SECURE_LOGGER.isDebugEnabled()) {
      SECURE_LOGGER.debug("Forskudd - resultat av beregning: {}", resultatFraCore.getBeregnetForskuddPeriodeListe());
    }

    var grunnlagReferanseListe = lagGrunnlagReferanseListe(grunnlag, resultatFraCore);
    return HttpResponse.Companion.from(HttpStatus.OK, new BeregnetForskuddResultat(resultatFraCore, grunnlagReferanseListe));
  }

  // Lager en liste over resultatgrunnlag som inneholder:
  //   - mottatte grunnlag som er brukt i beregningen
  //   - sjabloner som er brukt i beregningen
  private List<ResultatGrunnlag> lagGrunnlagReferanseListe(BeregnForskuddGrunnlag forskuddGrunnlag, BeregnetForskuddResultatCore resultatFraCore) {
    var mapper = new ObjectMapper();
    var resultatGrunnlagListe = new ArrayList<ResultatGrunnlag>();

    var grunnlagReferanseListe = resultatFraCore.getBeregnetForskuddPeriodeListe().stream()
        .map(ResultatPeriodeCore::getGrunnlagReferanseListe)
        .flatMap(Collection::stream)
        .distinct()
        .toList();

    // Matcher mottatte grunnlag med grunnlag som er brukt i beregningen
    resultatGrunnlagListe.addAll(forskuddGrunnlag.getGrunnlagListe().stream()
        .filter(grunnlag -> grunnlagReferanseListe.contains(grunnlag.getReferanse()))
        .map(grunnlag -> new ResultatGrunnlag(grunnlag.getReferanse(), grunnlag.getType(), grunnlag.getInnhold()))
        .toList());

    // Danner grunnlag basert på liste over sjabloner som er brukt i beregningen
    resultatGrunnlagListe.addAll(resultatFraCore.getSjablonListe().stream()
        .map(sjablon -> {
              var map = new LinkedHashMap<String, Object>();
              map.put("datoFom", mapDato(sjablon.getPeriode().getDatoFom()));
              map.put("datoTil", mapDato(sjablon.getPeriode().getDatoTil()));
              map.put("sjablonNavn", sjablon.getNavn());
              map.put("sjablonVerdi", sjablon.getVerdi().intValue());
              return new ResultatGrunnlag(sjablon.getReferanse(), "SJABLON", mapper.valueToTree(map));
            }
        )
        .toList());

    return resultatGrunnlagListe;
  }

  // Unngå å legge ut datoer høyere enn 9999-12-31
  private static String mapDato(LocalDate dato) {
    return dato.isAfter(LocalDate.parse("9999-12-31")) ? "9999-12-31" : dato.toString();
  }
}

File: src/main/kotlin/no/nav/bidrag/beregn/forskudd/rest/service/BeregnForskuddService.kt
package no.nav.bidrag.beregn.forskudd.rest.service

import com.fasterxml.jackson.databind.ObjectMapper
import no.nav.bidrag.beregn.forskudd.core.ForskuddCore
import no.nav.bidrag.beregn.forskudd.core.dto.BeregnetForskuddResultatCore
import no.nav.bidrag.beregn.forskudd.core.dto.ResultatPeriodeCore
import no.nav.bidrag.beregn.forskudd.rest.SECURE_LOGGER
import no.nav.bidrag.beregn.forskudd.rest.consumer.SjablonConsumer
import no.nav.bidrag.beregn.forskudd.rest.consumer.Sjablontall
import no.nav.bidrag.beregn.forskudd.rest.exception.UgyldigInputException
import no.nav.bidrag.commons.web.HttpResponse
import no.nav.bidrag.commons.web.HttpResponse.Companion.from
import no.nav.bidrag.domene.enums.Grunnlagstype
import no.nav.bidrag.domene.enums.resultatkoder.ResultatKodeForskudd
import no.nav.bidrag.domene.tid.ÅrMånedsperiode
import no.nav.bidrag.transport.behandling.beregning.felles.BeregnGrunnlag
import no.nav.bidrag.transport.behandling.beregning.felles.Grunnlag
import no.nav.bidrag.transport.behandling.beregning.felles.valider
import no.nav.bidrag.transport.behandling.beregning.forskudd.BeregnetForskuddResultat
import no.nav.bidrag.transport.behandling.beregning.forskudd.ResultatBeregning
import no.nav.bidrag.transport.behandling.beregning.forskudd.ResultatPeriode
import org.slf4j.LoggerFactory
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Service
import java.time.LocalDate

@Service
class BeregnForskuddService(private val sjablonConsumer: SjablonConsumer, private val forskuddCore: ForskuddCore) {

    fun beregn(grunnlag: BeregnGrunnlag): HttpResponse<BeregnetForskuddResultat> {
        if (SECURE_LOGGER.isDebugEnabled) {
            SECURE_LOGGER.debug("Mottatt følgende request: {}", grunnlag)
        }

        // Kontroll av inputdata
        try {
            grunnlag.valider()
        } catch (e: IllegalArgumentException) {
            throw UgyldigInputException("Ugyldig input ved beregning av forskudd: " + e.message)
        }

        // Henter sjabloner
        val sjablonSjablontallResponse: HttpResponse<List<Sjablontall>> = sjablonConsumer.hentSjablonSjablontall()

        if (sjablonSjablontallResponse.responseEntity.body.isNullOrEmpty()) {
            LOGGER.error("Klarte ikke å hente sjabloner")
            return from(httpStatus = HttpStatus.OK, body = BeregnetForskuddResultat())
        }

        val sjablonTallListe = sjablonSjablontallResponse.responseEntity.body!!

        if (LOGGER.isDebugEnabled) {
            LOGGER.debug("Antall sjabloner hentet av type Sjablontall: ${sjablonSjablontallResponse.responseEntity.body!!.size}")
        }

        // Lager input-grunnlag til core-modulen
        val grunnlagTilCore = CoreMapper.mapGrunnlagTilCore(
            beregnForskuddGrunnlag = grunnlag,
            sjablontallListe = sjablonTallListe
        )

        if (SECURE_LOGGER.isDebugEnabled) {
            SECURE_LOGGER.debug("Forskudd - grunnlag for beregning: {}", grunnlagTilCore)
        }

        // Kaller core-modulen for beregning av forskudd
        val resultatFraCore = try {
            forskuddCore.beregnForskudd(grunnlagTilCore)
        } catch (e: Exception) {
            throw UgyldigInputException("Ugyldig input ved beregning av forskudd: " + e.message)
        }

        if (resultatFraCore.avvikListe.isNotEmpty()) {
            val avvikTekst = resultatFraCore.avvikListe.joinToString("; ") { it.avvikTekst }
            LOGGER.warn("Ugyldig input ved beregning av forskudd. Følgende avvik ble funnet: $avvikTekst")
            SECURE_LOGGER.warn("Ugyldig input ved beregning av forskudd. Følgende avvik ble funnet: $avvikTekst")
            SECURE_LOGGER.info(
                "Forskudd - grunnlag for beregning: " + System.lineSeparator() +
                    "beregnDatoFra= " + grunnlagTilCore.beregnDatoFra + System.lineSeparator() +
                    "beregnDatoTil= " + grunnlagTilCore.beregnDatoTil + System.lineSeparator() +
                    "soknadBarn= " + grunnlagTilCore.soknadBarn + System.lineSeparator() +
                    "barnIHusstandenPeriodeListe= " + grunnlagTilCore.barnIHusstandenPeriodeListe + System.lineSeparator() +
                    "inntektPeriodeListe= " + grunnlagTilCore.inntektPeriodeListe + System.lineSeparator() +
                    "sivilstandPeriodeListe= " + grunnlagTilCore.sivilstandPeriodeListe + System.lineSeparator()
            )
            throw UgyldigInputException("Ugyldig input ved beregning av forskudd. Følgende avvik ble funnet: $avvikTekst")
        }

        if (SECURE_LOGGER.isDebugEnabled) {
            SECURE_LOGGER.debug("Forskudd - resultat av beregning: {}", resultatFraCore.beregnetForskuddPeriodeListe)
        }

        val grunnlagReferanseListe = lagGrunnlagReferanseListe(forskuddGrunnlag = grunnlag, resultatFraCore = resultatFraCore)

        val respons = BeregnetForskuddResultat(
            beregnetForskuddPeriodeListe = mapFraResultatPeriodeCore(resultatFraCore.beregnetForskuddPeriodeListe),
            grunnlagListe = grunnlagReferanseListe
        )

        if (SECURE_LOGGER.isDebugEnabled) {
            SECURE_LOGGER.debug("Returnerer følgende respons: {}", respons)
        }

        return from(httpStatus = HttpStatus.OK, body = respons)
    }

    private fun mapFraResultatPeriodeCore(resultatPeriodeCoreListe: List<ResultatPeriodeCore>) =
        resultatPeriodeCoreListe.map {
            ResultatPeriode(
                periode = ÅrMånedsperiode(fom = it.periode.datoFom, til = it.periode.datoTil),
                resultat = ResultatBeregning(
                    belop = it.resultat.belop,
                    kode = ResultatKodeForskudd.valueOf(it.resultat.kode),
                    regel = it.resultat.regel
                ),
                grunnlagReferanseListe = it.grunnlagReferanseListe
            )
        }

    // Lager en liste over resultatgrunnlag som inneholder:
    //   - mottatte grunnlag som er brukt i beregningen
    //   - sjabloner som er brukt i beregningen
    private fun lagGrunnlagReferanseListe(forskuddGrunnlag: BeregnGrunnlag, resultatFraCore: BeregnetForskuddResultatCore): List<Grunnlag> {
        val mapper = ObjectMapper()
        val resultatGrunnlagListe = mutableListOf<Grunnlag>()
        val grunnlagReferanseListe = resultatFraCore.beregnetForskuddPeriodeListe
            .flatMap { it.grunnlagReferanseListe }
            .distinct()

        // Matcher mottatte grunnlag med grunnlag som er brukt i beregningen
        resultatGrunnlagListe.addAll(
            forskuddGrunnlag.grunnlagListe!!
                .filter { grunnlagReferanseListe.contains(it.referanse) }
                .map { Grunnlag(referanse = it.referanse, type = it.type, innhold = it.innhold) }
        )

        // Danner grunnlag basert på liste over sjabloner som er brukt i beregningen
        resultatGrunnlagListe.addAll(
            resultatFraCore.sjablonListe
                .map {
                    val map = LinkedHashMap<String, Any>()
                    map["datoFom"] = mapDato(it.periode.datoFom)
                    map["datoTil"] = mapDato(it.periode.datoTil!!)
                    map["sjablonNavn"] = it.navn
                    map["sjablonVerdi"] = it.verdi.toInt()
                    Grunnlag(referanse = it.referanse, type = Grunnlagstype.SJABLON, innhold = mapper.valueToTree(map))
                }
        )

        return resultatGrunnlagListe
    }

    companion object {
        private val LOGGER = LoggerFactory.getLogger(BeregnForskuddService::class.java)

        // Unngå å legge ut datoer høyere enn 9999-12-31
        private fun mapDato(dato: LocalDate): String {
            return if (dato.isAfter(LocalDate.parse("9999-12-31"))) "9999-12-31" else dato.toString()
        }
    }
}
